Nama : Tegas Gagas Impian Prahita N

In [1]:
import pandas as pd

# Soal Beginner

## No.1

In [2]:
# Masukan syntax jawaban di sini

data = pd.DataFrame({'no'    : [1, 2, 3, 4, 5],
                     'nama_siswa'  : ['Adi', 'Budi', 'Carli', 'Dadan', 'Endang'],
                     'umur': [15, 16, 14, None, 16],
                     'email'  : ['andi@gmail.com', 'budi@gmail.com', 'carli@gmail.com', 'dadan@gmail.com', 'endang@gmail.com'],
                     'score_matematika'  : [80, 90, 100, 95, 97],
                     'score_ekonomi'  : [90, 95, 100, 86, None],
                     'score_olahraga'  : [97, 95, 96, 100, 94]
                    })

data

no nama_siswa  umur             email  score_matematika  score_ekonomi  \
0   1        Adi  15.0    andi@gmail.com                80           90.0   
1   2       Budi  16.0    budi@gmail.com                90           95.0   
2   3      Carli  14.0   carli@gmail.com               100          100.0   
3   4      Dadan   NaN   dadan@gmail.com                95           86.0   
4   5     Endang  16.0  endang@gmail.com                97            NaN   

   score_olahraga  
0              97  
1              95  
2              96  
3             100  
4              94

## No.2 

In [3]:
# Masukan sytnax jawaban di sini

data_sorted = data.sort_values('score_matematika', ascending=False)

data_lowest_score = data_sorted.tail(1)

print(f"Nama siswa dengan score_matematika terkecil adalah {data_lowest_score.at[0,'nama_siswa']}")

Nama siswa dengan score_matematika terkecil adalah Adi


## No.3

In [4]:
# Masukan syntax jawaban di sini

data_siswa_gmail = data[data["email"].str.contains("gmail") & data["score_ekonomi"].notnull()]

umur_mean = data_siswa_gmail["umur"].mean()

print(f"Umur rata-rata dari siswa di 'data_siswa_gmail' adalah {umur_mean}")

Umur rata-rata dari siswa di 'data_siswa_gmail' adalah 15.0


## No.4

In [5]:
# Masukan syntax jawaban di sini

additional_data = pd.DataFrame({'no'    : [6, 7],
                     'nama_siswa'  : ['Fauzan', 'Guntur'],
                     'umur': [15, 16],
                     'email'  : ['fauzan@gmail.com', 'guntur@gmail.com'],
                     'score_matematika'  : [80, 90],
                     'score_ekonomi'  : [90, 95],
                     'score_olahraga'  : [97, 95]
                    })

# new_data = data.append(additional_data, ignore_index=True)
new_data = pd.concat([data, additional_data], ignore_index=True)
# print(new_data)

umur_med = new_data["umur"].median()

print(f"Median umur mereka sekarang adalah {umur_med}")

Median umur mereka sekarang adalah 15.5


## No.5

In [6]:
##### Masukan syntax jawaban di sini

data_address = pd.DataFrame({'nama'    : ['Adi', 'Budi', 'Dadan', 'Endang', 'Guntur', 'Halim'],
                     'alamat'  : ['DKI Jakarta', 'DKI Jakarta', 'Banten', 'Banten', 'Jawa Barat', 'Aceh']
                    })


# merge data frame “new_data”
data_merge = new_data.merge(data_address, left_on='nama_siswa', right_on='nama', how='outer')

# umur rata-rata siswa per alamat mereka berasal
umur_add_mean = data_merge.groupby('alamat').agg({'umur' : 'mean'})

result = umur_add_mean.loc[umur_add_mean['umur'] == umur_add_mean['umur'].min()]

print(f"Umur rata-rata mereka yang paling kecil pada alamat  {result}")

Umur rata-rata mereka yang paling kecil pada alamat               umur
alamat           
DKI Jakarta  15.5


# Soal Intermediate

## No.6

In [7]:
# Masukan syntax jawaban di sini

airlines = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Vemby\data_airlines.xlsx')
airports = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Vemby\data_airports.xlsx')
flights = pd.read_excel(r'C:\Users\tegas\Documents\rakamin\Vemby\data_flights.xlsx')

In [8]:
flights.head()

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0  2015      1    1            4      UA           1475      N69819   
1  2015      1    1            4      AA            227      N571AA   
2  2015      1    1            4      AA            357      N3HAAA   
3  2015      1    1            4      MQ           3459      N671MQ   
4  2015      1    1            4      OO           6559      N886AS   

  ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  ARRIVAL_TIME  \
0            ORD                 MCO                 1600  ...        2146.0   
1            DFW                 MSP                 2040  ...        2321.0   
2            BWI                 MIA                 1726  ...        2003.0   
3            MIA                 PNS                 1635  ...        1751.0   
4            PHX                 PSP                 1943  ...        1946.0   

   ARRIVAL_DELAY  DIVERTED  CANCELLED  CANCELLATION_REASON  AIR_SYSTEM_DELAY  \
0          121.0         0          0                  NaN               0.0   
1           26.0         0          0                  NaN               0.0   
2           -9.0         0          0                  NaN               NaN   
3           27.0         0          0                  NaN               6.0   
4           -5.0         0          0                  NaN               NaN   

   SECURITY_DELAY  AIRLINE_DELAY  LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0             0.0           93.0                 28.0            0.0  
1             0.0           26.0                  0.0            0.0  
2             NaN            NaN                  NaN            NaN  
3             0.0           21.0                  0.0            0.0  
4             NaN            NaN                  NaN            NaN  

[5 rows x 31 columns]

In [9]:
flights['FLIGHT_NUMBER'].value_counts()

894     12
608     11
1373    11
679     10
233     10
        ..
2530     1
6515     1
4531     1
5146     1
5119     1
Name: FLIGHT_NUMBER, Length: 4314, dtype: int64

In [10]:
f = flights.groupby('MONTH').agg({'FLIGHT_NUMBER' : 'count',
                                    'CANCELLED' : 'sum',
                                    'ELAPSED_TIME' : ['mean', 'min', 'max']})
fly = f.reset_index()

fly.columns = ['bulan','total_penerbangan','total_pembatalan','waktu_tempuh_rerata','waktu_tempuh_min','waktu_tempuh_max']

fly

bulan  total_penerbangan  total_pembatalan  waktu_tempuh_rerata  \
0       1                798                20           132.153548   
1       2                715                30           140.375731   
2       3                877                22           142.357562   
3       4                824                 9           138.420859   
4       5                840                 5           136.516827   
5       6                839                18           139.057317   
6       7                937                14           141.110990   
7       8                887                 8           142.948689   
8       9                787                 1           137.667516   
9      10                856                 2           134.366940   
10     11                822                 8           139.198032   
11     12                818                17           144.340426   

    waktu_tempuh_min  waktu_tempuh_max  
0               31.0             428.0  
1               24.0             408.0  
2               30.0             622.0  
3               31.0             570.0  
4               34.0             466.0  
5               32.0             398.0  
6               30.0             511.0  
7               27.0             655.0  
8               34.0             495.0  
9               31.0             388.0  
10              32.0             392.0  
11              33.0             453.0

**Insight**

Selama tahun 2015, pada bulan Juli memiliki jadwal penerbangan tersibuk yaitu 937 dan bulan Februari memiliki jadwal penerbangan paling sedikit yaitu 715. Ketika bulan Februari juga terdapat pembatalan penerbangan yang paling banyak dibandingkan bulan lainnya. Bulan September hampir seluruh penerbangan diberangkatkan dan hanya terdapat 1 pembatalan saja.

Penerbangan pada bulan Juli merupakan yang teramai dikarenakan puncak libur musim panas sehingga customer banyak melakukan berpergian. Pada bulan Januari - Maret memiliki jumlah pembatalan penerbangan yang relatif lebih besar dibandingkan bulan lainnya (diatas 20 kali). Hal ini dimungkinkan karena adanya cuaca buruk saat penerbangan di musim dingin.

Waktu tempuh penerbangan (elapsed time) secara umum konstan untuk setiap bulan berkisar 140 menit, dengan waktu terpendek berkisar 30 menit dan waktu terpanjang berkisar 380 hingga 650 menit. Waktu tempuh yang bervariasi bergantung terhadap beberapa kodisi seperti cuaca dan tingkat kepadatan bandara.

## No.7

In [11]:
# Masukan syntax jawaban di sini
airports[airports["AIRPORT"].isnull()]

airports[airports["IATA_CODE"].astype(str).str.contains("JFK")].head()

airports[airports["IATA_CODE"].astype(str).str.contains("ECP")].head()


IATA_CODE                                          AIRPORT         CITY  \
96       ECP  Northwest Florida Beaches International Airport  Panama City   

   STATE COUNTRY  LATITUDE  LONGITUDE  
96    FL     USA       NaN        NaN

In [12]:
a = flights.merge(airports, left_on='ORIGIN_AIRPORT',right_on='IATA_CODE', how='left')
b = a.merge(airports, left_on='DESTINATION_AIRPORT',right_on='IATA_CODE', how='left')

b = b.rename(columns = {'AIRPORT_x' : 'ORI_AIRPORT', 
                        'AIRPORT_y' : 'DEST_AIRPORT'})

b[['AIRLINE', 'FLIGHT_NUMBER', 'ORI_AIRPORT', 'DEST_AIRPORT']]

AIRLINE  FLIGHT_NUMBER  \
0         UA           1475   
1         AA            227   
2         AA            357   
3         MQ           3459   
4         OO           6559   
...      ...            ...   
9995      AS            579   
9996      DL           2172   
9997      NK            801   
9998      EV           4093   
9999      WN           5119   

                                            ORI_AIRPORT  \
0                  Chicago O'Hare International Airport   
1               Dallas/Fort Worth International Airport   
2            Baltimore-Washington International Airport   
3                           Miami International Airport   
4              Phoenix Sky Harbor International Airport   
...                                                 ...   
9995  San Diego International Airport (Lindbergh Field)   
9996                        Tampa International Airport   
9997                       Detroit Metropolitan Airport   
9998         Savannah/Hilton Head International Airport   
9999                     Portland International Airport   

                                           DEST_AIRPORT  
0                         Orlando International Airport  
1          Minneapolis-Saint Paul International Airport  
2                           Miami International Airport  
3     Pensacola International Airport (Pensacola Gul...  
4                    Palm Springs International Airport  
...                                                 ...  
9995                     Portland International Airport  
9996   Hartsfield-Jackson Atlanta International Airport  
9997                      Orlando International Airport  
9998               Chicago O'Hare International Airport  
9999         John Wayne Airport (Orange County Airport)  

[10000 rows x 4 columns]

In [13]:
b.duplicated().any()

False

In [14]:
c = b[b['ORI_AIRPORT'].isin(['John F. Kennedy International Airport (New York International Airport)',
                             'Northwest Florida Beaches International Airport'])]

c[['AIRLINE', 'FLIGHT_NUMBER', 'ORI_AIRPORT', 'DEST_AIRPORT']]

AIRLINE  FLIGHT_NUMBER  \
44        AA           1275   
50        B6            353   
86        UA            445   
163       AA              1   
235       MQ           3378   
...      ...            ...   
9808      AA             19   
9896      DL            435   
9902      B6           1983   
9903      B6            518   
9952      AA           2445   

                                            ORI_AIRPORT  \
44    John F. Kennedy International Airport (New Yor...   
50    John F. Kennedy International Airport (New Yor...   
86    John F. Kennedy International Airport (New Yor...   
163   John F. Kennedy International Airport (New Yor...   
235   John F. Kennedy International Airport (New Yor...   
...                                                 ...   
9808  John F. Kennedy International Airport (New Yor...   
9896  John F. Kennedy International Airport (New Yor...   
9902  John F. Kennedy International Airport (New Yor...   
9903  John F. Kennedy International Airport (New Yor...   
9952  John F. Kennedy International Airport (New Yor...   

                                          DEST_AIRPORT  
44                               Cyril E. King Airport  
50                    Palm Beach International Airport  
86                   Los Angeles International Airport  
163                  Los Angeles International Airport  
235          Ronald Reagan Washington National Airport  
...                                                ...  
9808                 Los Angeles International Airport  
9896               San Francisco International Airport  
9902                     Orlando International Airport  
9903  Gen. Edward Lawrence Logan International Airport  
9952                     Orlando International Airport  

[181 rows x 4 columns]

In [15]:
d = c.groupby(['ORI_AIRPORT','DEST_AIRPORT']).agg({'FLIGHT_NUMBER' : 'count',
                                                   'DISTANCE' : 'mean'})

e = d.reset_index()

e.columns = ['asal','tujuan','total_penerbangan','jarak_rerata']

e

asal  \
0   John F. Kennedy International Airport (New Yor...   
1   John F. Kennedy International Airport (New Yor...   
2   John F. Kennedy International Airport (New Yor...   
3   John F. Kennedy International Airport (New Yor...   
4   John F. Kennedy International Airport (New Yor...   
5   John F. Kennedy International Airport (New Yor...   
6   John F. Kennedy International Airport (New Yor...   
7   John F. Kennedy International Airport (New Yor...   
8   John F. Kennedy International Airport (New Yor...   
9   John F. Kennedy International Airport (New Yor...   
10  John F. Kennedy International Airport (New Yor...   
11  John F. Kennedy International Airport (New Yor...   
12  John F. Kennedy International Airport (New Yor...   
13  John F. Kennedy International Airport (New Yor...   
14  John F. Kennedy International Airport (New Yor...   
15  John F. Kennedy International Airport (New Yor...   
16  John F. Kennedy International Airport (New Yor...   
17  John F. Kennedy International Airport (New Yor...   
18  John F. Kennedy International Airport (New Yor...   
19  John F. Kennedy International Airport (New Yor...   
20  John F. Kennedy International Airport (New Yor...   
21  John F. Kennedy International Airport (New Yor...   
22  John F. Kennedy International Airport (New Yor...   
23  John F. Kennedy International Airport (New Yor...   
24  John F. Kennedy International Airport (New Yor...   
25  John F. Kennedy International Airport (New Yor...   
26  John F. Kennedy International Airport (New Yor...   
27  John F. Kennedy International Airport (New Yor...   
28  John F. Kennedy International Airport (New Yor...   
29  John F. Kennedy International Airport (New Yor...   
30  John F. Kennedy International Airport (New Yor...   
31  John F. Kennedy International Airport (New Yor...   
32  John F. Kennedy International Airport (New Yor...   
33  John F. Kennedy International Airport (New Yor...   
34  John F. Kennedy International Airport (New Yor...   
35  John F. Kennedy International Airport (New Yor...   
36  John F. Kennedy International Airport (New Yor...   
37  John F. Kennedy International Airport (New Yor...   
38  John F. Kennedy International Airport (New Yor...   
39  John F. Kennedy International Airport (New Yor...   
40  John F. Kennedy International Airport (New Yor...   
41    Northwest Florida Beaches International Airport   
42    Northwest Florida Beaches International Airport   
43    Northwest Florida Beaches International Airport   
44    Northwest Florida Beaches International Airport   
45    Northwest Florida Beaches International Airport   

                                               tujuan  total_penerbangan  \
0              Austin-Bergstrom International Airport                  3   
1               Buffalo Niagara International Airport                  3   
2                    Burlington International Airport                  2   
3     Charleston International Airport/Charleston AFB                  5   
4             Charlotte Douglas International Airport                  8   
5                Chicago O'Hare International Airport                  2   
6             Cleveland Hopkins International Airport                  3   
7                               Cyril E. King Airport                  2   
8                        Denver International Airport                  2   
9     Fort Lauderdale-Hollywood International Airport                  6   
10   Gen. Edward Lawrence Logan International Airport                 15   
11            Greater Rochester International Airport                  1   
12   Hartsfield-Jackson Atlanta International Airport                  3   
13                     Honolulu International Airport                  1   
14                 Indianapolis International Airport                  1   
15                 Jacksonville International Airport                  3   
16               Long Beach Airport (Daugherty Field)                  1   
1

In [16]:
f = pd.pivot_table(e,
                   index = ['tujuan'],
                   columns = ['asal'],
                   values = ['total_penerbangan','jarak_rerata']
                  )
f

jarak_rerata  \
asal                                              John F. Kennedy International Airport (New York International Airport)   
tujuan                                                                                                                     
Austin-Bergstrom International Airport                                                        1521.0                       
Buffalo Niagara International Airport                                                          301.0                       
Burlington International Airport                                                               266.0                       
Charleston International Airport/Charleston AFB                                                636.0                       
Charlotte Douglas International Airport                                                        541.0                       
Chicago O'Hare International Airport                                                           740.0                       
Cleveland Hopkins International Airport                                                        425.0                       
Cyril E. King Airport                                                                         1623.0                       
Dallas Love Field                                                                                NaN                       
Denver International Airport                                                                  1626.0                       
Fort Lauderdale-Hollywood International Airport                                               1069.0                       
Gen. Edward Lawrence Logan International Airport                                               187.0                       
George Bush Intercontinental Airport                                                             NaN                       
Greater Rochester International Airport                                                        264.0                       
Hartsfield-Jackson Atlanta International Airport                                               760.0                       
Honolulu International Airport                                                                4983.0                       
Indianapolis International Airport                                                             665.0                       
Jacksonville International Airport                                                             828.0                       
Long Beach Airport (Daugherty Field)                                                          2465.0                       
Los Angeles International Airport                                                             2475.0                       
Louis Armstrong New Orleans International Airport                                             1182.0                       
Luis Muñoz Marín International Airport                                                        1598.0                       
McCarran International Airport                                                                2248.0                       
Mercedita Airport                                                                             1617.0                       
Miami International Airport                                                                   1089.0                       
Nashville International Airport                                                                765.0                       
Oakland International Airport                                                                 2576.0                       
Orlando International Airport                                                                  944.0                       
Palm Beach International Airport                                                              1028.0                       
Phoenix Sky Harbor International Airport                                                      2153.0                       
Port Columbus In

Destinasi airport mana yang paling banyak jumlah penerbangannya yaitu:

| Asal | Tujuan | Jumlah | Jarak tempuh |
|---|---|---|---|
| John F. Kennedy | San Francisco | 23 | 2586.0 |
| Northwest Florida Beaches | Hartsfield-Jackson Atlanta | 3 | 240.0 |

# Soal Advanced

## No.8

In [17]:
# Masukan syntax jawaban di sini

df_a = pd.read_csv(r'C:\Users\tegas\Documents\rakamin\Vemby\01 telecom_revenue.csv')
df_b = pd.read_csv(r'C:\Users\tegas\Documents\rakamin\Vemby\02 telecom_usage.csv')

df_1 = df_a[(df_a['Occupation'].isin(['Professional', 'Student', 'Crafts'])) & (df_a['MonthlyRevenue'] > 10)] 
df_2 = df_b[(df_b['UnansweredCalls'] > df_b['BlockedCalls']) & 
             ((df_b["CustomerCareCalls"].astype(str).str.contains("^8")) | 
              (df_b["CustomerCareCalls"].astype(str).str.contains("^4")) |
              (df_b["CustomerCareCalls"].astype(str).str.contains("^2")))]  

In [18]:
df_1

CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
12       3397170    No           83.51           290.0                  93.0   
25       3274822    No           30.83           273.0                  40.0   
26       3394258    No           52.72          1501.0                  62.0   
29       3056178   Yes           54.09           412.0                  40.0   
35       3065718    No           43.06           110.0                  35.0   
...          ...   ...             ...             ...                   ...   
9964     3015822    No           46.10           203.0                  38.0   
9968     3323878    No           44.26           419.0                  45.0   
9970     3050290   Yes           51.06           192.0                  30.0   
9981     3119258    No           39.49           352.0                  45.0   
9994     3125862    No           22.50            18.0                  17.0   

      OverageMinutes  RoamingCalls    Occupation  
12               0.0           0.0  Professional  
25               0.0           0.8  Professional  
26               0.0           0.0        Crafts  
29              46.0           0.0  Professional  
35               0.0           3.3        Crafts  
...              ...           ...           ...  
9964            30.0           0.0        Crafts  
9968             2.0           0.0  Professional  
9970            74.0           0.0  Professional  
9981             0.0           1.7        Crafts  
9994             2.0           0.0  Professional  

[2023 rows x 8 columns]

In [19]:
df_2

CustomerID  DroppedCalls  BlockedCalls  UnansweredCalls  \
7        3357910          15.7          27.0             91.0   
20       3089434           6.7           0.7             26.3   
25       3187322           2.0           0.0             24.3   
34       3124478          22.0           0.3              3.3   
39       3226290           9.7           3.7             32.7   
...          ...           ...           ...              ...   
4919     3349482           9.7           5.0             56.0   
4930     3378670          15.3           1.3            154.0   
4946     3147650           2.7           0.7             25.7   
4948     3299922           4.3           4.3             29.3   
4960     3076294           3.0           0.0             29.3   

      CustomerCareCalls  
7                   8.0  
20                  2.3  
25                  2.7  
34                  8.7  
39                  2.0  
...                 ...  
4919                2.0  
4930               23.7  
4946                2.7  
4948                2.3  
4960                2.0  

[495 rows x 5 columns]

In [20]:
a = df_1.merge(df_2, on='CustomerID')

a.sort_values('BlockedCalls', ascending=False)

CustomerID Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
2      3338506    No           73.47           900.0                  50.0   
13     3006690    No           84.77           827.0                  70.0   
9      3313586    No           36.12           287.0                  45.0   
15     3303050   Yes          120.39          2524.0                  85.0   
14     3179170    No          118.44          1447.0                  85.0   
8      3175862    No           88.11           852.0                  41.0   
12     3235462    No           31.00           407.0                  40.0   
10     3055270    No          100.81          2233.0                  85.0   
16     3248414   Yes           59.88           244.0                  68.0   
0      3213958   Yes           34.74           228.0                  41.0   
7      3356414    No           56.81           953.0                  70.0   
5      3227662    No           53.24           703.0                  70.0   
1      3326638    No           54.18           734.0                  70.0   
11     3031378    No           56.98            44.0                  70.0   
6      3278010    No           40.73           455.0                  40.0   
4      3008986    No           34.54            40.0                  34.0   
3      3104186    No           39.70           451.0                  30.0   
17     3365526   Yes           64.46            63.0                  80.0   

    OverageMinutes  RoamingCalls    Occupation  DroppedCalls  BlockedCalls  \
2            100.0           0.0  Professional           7.7          56.0   
13            76.0           0.5  Professional           4.7          13.3   
9              0.0           0.2  Professional           1.3          10.3   
15           155.0           0.0  Professional           7.0           9.3   
14            91.0           3.0        Crafts          12.7           9.0   
8            110.0           0.0        Crafts           8.7           4.7   
12             0.0           0.0  Professional           1.3           3.3   
10            82.0           0.0  Professional          12.0           2.3   
16            10.0           0.0  Professional           1.0           1.3   
0              0.0           0.0  Professional           2.0           1.3   
7             13.0           0.0  Professional          19.7           0.7   
5              0.0           0.0  Professional           3.7           0.7   
1              0.0           0.0        Crafts           8.7           0.3   
11             0.0           0.0  Professional           1.0           0.0   
6              0.0           0.0  Professional           2.7           0.0   
4              6.0           0.0  Professional           1.7           0.0   
3              2.0           0.0  Professional           5.7           0.0   
17             0.0           0.5  Professional           0.3           0.0   

    UnansweredCalls  CustomerCareCalls  
2              58.0                2.3  
13             42.3                2.3  
9              11.0                4.7  
15             95.3                2.3  
14            110.3               29.3  
8              44.7                4.0  
12              4.0                2.3  
10            125.7                8.7  
16             17.7                4.3  
0              21.0                4.3  
7              62.0                4.7  
5              40.7                2.0  
1              29.0                2.0  
11              1.0                2.0  
6              38.3                4.7  
4               2.3                2.0  
3               7.7                2.7  
17              7.7                2.7

In [21]:
b = a.groupby(['Occupation']).agg({'CustomerID' : 'nunique',
                                   'DroppedCalls' : 'mean'})
b

CustomerID  DroppedCalls
Occupation                            
Crafts                 3     10.033333
Professional          15      4.786667

Rata-rata DroppedCalls paling besar terdapat pada pekerjaan Crafts